hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1066,17 +1097,49 @@ private boolean validateTimelineBeforeSchedulingCompaction(Option<String> inFlig
     // c. Do consider out of order commits. For eg, c4 from DT could complete before c3. and we can't trigger compaction in MDT with c4 as base instant time, until every
     // instant before c4 is synced with metadata table.
-    List<HoodieInstant> pendingInstants = dataMetaClient.reloadActiveTimeline().filterInflightsAndRequested()
-        .findInstantsBeforeOrEquals(latestDeltaCommitTimeInMetadataTable).getInstants();
 
-    if (!pendingInstants.isEmpty()) {
-      checkNumDeltaCommits(metadataMetaClient, dataWriteConfig.getMetadataConfig().getMaxNumDeltacommitsWhenPending());
+    // There should not be any incomplete instants on MDT
+    HoodieActiveTimeline metadataTimeline = metadataMetaClient.reloadActiveTimeline();
+    List<HoodieInstant> pendingInstantsOnMetadataTable = metadataTimeline.filterInflightsAndRequested()
+        .getInstants();
+    if (!pendingInstantsOnMetadataTable.isEmpty()) {
       LOG.info(String.format(
-          "Cannot compact metadata table as there are %d inflight instants in data table before latest deltacommit in metadata table: %s. Inflight instants in data table: %s",
-          pendingInstants.size(), latestDeltaCommitTimeInMetadataTable, Arrays.toString(pendingInstants.toArray())));
+          "Cannot compact MDT as there are %d inflight instants: %s",
+          pendingInstantsOnMetadataTable.size(), Arrays.toString(pendingInstantsOnMetadataTable.toArray())));
+      metrics.ifPresent(m -> m.incrementMetric(HoodieMetadataMetrics.SKIP_TABLE_SERVICES, 1));
       return false;
     }
 
+    // There should not be any incomplete instants on dataset
+    HoodieActiveTimeline datasetTimeline = dataMetaClient.reloadActiveTimeline();
+    List<HoodieInstant> pendingInstantsOnDataset = datasetTimeline.filterInflightsAndRequested().getInstantsAsStream()
+        .filter(i -> !inFlightInstantTimestamp.isPresent() || !i.getTimestamp().equals(inFlightInstantTimestamp.get()))
+        .collect(Collectors.toList());
+    if (!pendingInstantsOnDataset.isEmpty()) {
+      LOG.info(String.format(
+          "Cannot compact MDT as there are %d inflight instants on dataset before latest deltacommit %s: %s",
+          pendingInstantsOnDataset.size(), latestDeltaCommitTimeInMetadataTable,
+          Arrays.toString(pendingInstantsOnDataset.toArray())));
+      metrics.ifPresent(m -> m.incrementMetric(HoodieMetadataMetrics.SKIP_TABLE_SERVICES, 1));
+      return false;
+    }
+
+    // Check if the inflight commit is greater than all the completed commits.
+    Option<HoodieInstant> lastCompletedInstant = dataMetaClient.getActiveTimeline().filterCompletedInstants().lastInstant();
+    if (!lastCompletedInstant.isPresent()) {
+      LOG.info("Last completed commit is not present.");
+      metrics.ifPresent(m -> m.incrementMetric(HoodieMetadataMetrics.SKIP_TABLE_SERVICES, 1));
+      return false;
+    }
+    if (HoodieTimeline.compareTimestamps(lastCompletedInstant.get().getTimestamp(),
```

Comment:
may I know why we need this specifically. we should be able to allow out of order commits. All we need to ensure is, the new compaction in MDT should have an instant time where in all commits prior to that is synced to MDT.